### PR TITLE
update "OrderViewSet" class

### PR DESCRIPTION
**Pull request by ahsan-ul-haq**
On staff users can filter orders by username of any user.
ECOM-4550

### CODE REVIEW

**Comment by ahsan-ul-haq** [comment]
@clintonb please review

**Comment by clintonb** [comment]
I'd much rather see a username querystring filter (e.g. https://ecommerce.example.com/api/v2/orders/?username=clintonb). DRF has support for this: http://www.django-rest-framework.org/api-guide/filtering/#djangofilterbackend.

**Comment by tasawernawaz** [comment]
@clintonb I have add django filter but have some question.
Previously all users can see there own orders. 
After adding filter, user can filter like this `/api/v2/orders?username=<username>`, but that filter is applied only on queryset overridden [here](https://github.com/edx/ecommerce/pull/759/files#diff-eea0521d5c26662ce10e8a69057642b0L27) which is already filtering queryset on requested user.. Please let me know what is expected behavior ?

**Comment by ahsan-ul-haq** [comment]
I would suggest that leave [this](https://github.com/edx/ecommerce/pull/759/files#diff-eea0521d5c26662ce10e8a69057642b0L27) as it was before and add django filter on the resultant data.

**Comment by tasawernawaz** [comment]
Updated code as per your suggestion. @ahsan-ul-haq

File: ecommerce/extensions/api/v2/tests/views/test_orders.py [hunk 1]
```diff
@@ -95,9 +95,10 @@ def test_with_other_users_orders(self):
         (True, False),
     )
     def test_staff_superuser(self, is_staff, is_superuser):
-        """ The view should return all orders for when authenticating as a staff member or superuser. """
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_orders.py:98]: Revert.

**Comment by clintonb** [comment]
@tasawernawaz your new code needs tests.

The logic for this view should be as follows:
1. If the user is not a staff user, the user only sees his/her own orders regardless of the username filter. It would be preferable to return a 403 status code informing the user he is not allowed to filter by the username of other users as a non-staff user.
2. If the user is a staff user, by default (e.g. no filter) he sees all orders. Adding a filter should limit the orders to those of the requested user.

**Comment by tasawernawaz** [comment]
@clintonb Made suggested changes in view. Please review.

File: ecommerce/extensions/api/v2/views/orders.py [hunk 2]
```diff
@@ -23,16 +25,21 @@ class OrderViewSet(viewsets.ReadOnlyModelViewSet):
     queryset = Order.objects.all()
     serializer_class = serializers.OrderSerializer
     throttle_classes = (ServiceUserThrottle,)
+    filter_backends = (filters.DjangoFilterBackend,)
+    filter_class = OrderFilter
 
-    def filter_queryset(self, queryset):
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views/orders.py:27]: Why was this changed?

File: ecommerce/extensions/api/v2/views/orders.py [hunk 3]
```diff
@@ -23,16 +25,21 @@ class OrderViewSet(viewsets.ReadOnlyModelViewSet):
...
-        queryset = super(OrderViewSet, self).filter_queryset(queryset)
+    def list(self, request, *args, **kwargs):
+        username = self.request.query_params.get('username')
         user = self.request.user
 
-        # Non-staff users should only see their own orders
         if not user.is_staff:
-            queryset = queryset.filter(user=user)
+            if username and user.username != username:
+                return Response("You are not allowed to filter on any other username.",
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views/orders.py]: `raise PermissionDenied`

File: ecommerce/extensions/api/v2/views/orders.py (hunk 2 above)
Comment (tasawernawaz) [inline ecommerce/extensions/api/v2/views/orders.py:27]: I have override `List` view. Do you think this method still required ?

Comment (clintonb) [inline ecommerce/extensions/api/v2/views/orders.py:27]: Override `filter_queryset`.

**Comment by tasawernawaz** [comment]
@clintonb Made suggested change. Please review.

File: ecommerce/extensions/api/v2/views/orders.py [hunk 4]
```diff
@@ -23,13 +26,19 @@ class OrderViewSet(viewsets.ReadOnlyModelViewSet):
...
+    filter_backends = (filters.DjangoFilterBackend,)
+    filter_class = OrderFilter
 
     def filter_queryset(self, queryset):
         queryset = super(OrderViewSet, self).filter_queryset(queryset)
+
+        username = self.request.query_params.get('username')
         user = self.request.user
 
-        # Non-staff users should only see their own orders
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views/orders.py:31]: This comment is still relevant.

File: ecommerce/extensions/api/v2/views/orders.py [hunk 5]
```diff
@@ -1,17 +1,20 @@
...
+from rest_framework.exceptions import PermissionDenied
 
 from oscar.core.loading import get_model, get_class
-from rest_framework import status, viewsets
+from rest_framework import filters, status, viewsets
 from rest_framework.decorators import detail_route
 from rest_framework.permissions import IsAuthenticated, DjangoModelPermissions
 from rest_framework.response import Response
 
+from ecommerce.extensions.api.filters import OrderFilter
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/views/orders.py]: Alphabetize.

File: ecommerce/extensions/api/v2/tests/views/test_orders.py [hunk 6]
```diff
@@ -116,6 +116,37 @@ def test_user_information(self):
...
+        """ Verify the "staff" user can filter data by username. Non staff
+        users are not allowed to filter on any other username."""
+
+        # create two orders for different users
+        order = factories.create_order(user=self.user)
+        other_user = self.create_user()
+        other_order = factories.create_order(user=other_user)
+
+        requester = self.create_user(is_staff=is_staff)
+        token = self.generate_jwt_token_header(requester)
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_orders.py]: You don't need a token. Call `self.client.login`.

File: ecommerce/extensions/api/v2/tests/views/test_orders.py [hunk 7]
```diff
@@ -116,6 +116,37 @@ def test_user_information(self):
         self.assertEqual(content['results'][0]['user']['email'], admin_user.email)
         self.assertEqual(content['results'][0]['user']['username'], admin_user.username)
 
+    @ddt.unpack
+    @ddt.data(
+        {'is_staff': True, 'status_code': 200},
+        {'is_staff': False, 'status_code': 403},
+    )
+    def test_username_filter(self, is_staff, status_code):
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_orders.py]: Split this into two tests. You can share code with helper methods, but this will be a lot cleaner as two separate tests.

File: ecommerce/extensions/api/v2/tests/views/test_orders.py [hunk 8]
```diff
@@ -116,6 +116,37 @@ def test_user_information(self):
...
+        token = self.generate_jwt_token_header(requester)
+
+        self.assert_list_with_username_filter(self.user, order, token, status_code)
+        self.assert_list_with_username_filter(other_user, other_order, token, status_code)
+
+    def assert_list_with_username_filter(self, user, order, token, expected_status_code):
+        response = self.client.get(self.path, data={'username': user.username}, HTTP_AUTHORIZATION=token)
+        self.assertEqual(response.status_code, expected_status_code)
+
+        if expected_status_code == 200:
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_orders.py]: Assertion methods shouldn't have if statements. This is a sign that you should have a separate assertion method.

File: ecommerce/extensions/api/v2/tests/views/test_orders.py [hunk 9]
```diff
@@ -116,6 +116,37 @@ def test_user_information(self):
...
+            content = json.loads(response.content)
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_orders.py]: Use the serializer and look at `response.data`.

File: ecommerce/extensions/api/filters.py [hunk 10]
```diff
@@ -18,3 +19,13 @@ def filter_product_class(self, queryset, value):
     class Meta(object):
         model = Product
         fields = ('product_class', 'structure', 'title',)
+
+
+class OrderFilter(django_filters.FilterSet):
+    """ Filter orders via query string parameter."""
+
+    username = django_filters.CharFilter(name="user__username")
```
Comment (zubair-arbi) [inline ecommerce/extensions/api/filters.py]: use single quotes

**Comment by tasawernawaz** [comment]
@clintonb Please review.

File: ecommerce/extensions/api/v2/tests/views/test_orders.py [hunk 11]
```diff
@@ -8,7 +8,9 @@
 from django.test import override_settings
 from oscar.core.loading import get_model
 from oscar.test import factories
+from threadlocals.threadlocals import get_current_request
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_orders.py]: Get rid of this and avoid using it. It will be going away soon. Use RequestFactory.

File: ecommerce/extensions/api/v2/tests/views/test_orders.py [hunk 12]
```diff
@@ -116,6 +118,37 @@ def test_user_information(self):
         self.assertEqual(content['results'][0]['user']['email'], admin_user.email)
         self.assertEqual(content['results'][0]['user']['username'], admin_user.username)
 
+    def test_username_filter_with_staff(self):
+        """ Verify the "staff" user can filter data by username."""
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_orders.py]: Staff doesn't need to be in quotes.

File: ecommerce/extensions/api/v2/tests/views/test_orders.py [hunk 13]
```diff
@@ -116,6 +118,37 @@ def test_user_information(self):
...
+        requester = self.create_user(is_staff=False)
+        self.client.login(username=requester.username, password=self.password)
+
+        response = self.client.get(self.path, data={'username': self.user.username})
+        self.assertEqual(response.status_code, 403)
+
+    def assert_list_with_username_filter(self, user, order):
+        """ Helper method for making assertions. """
+
+        response = self.client.get(self.path, data={'username': user.username})
```
Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_orders.py]: Passing a data kwarg is now how you set a query string.

**Comment by clintonb** [comment]
:+1: (once remaining comments are addressed)

File: ecommerce/extensions/api/v2/tests/views/test_orders.py (hunk 11 above)
Comment (tasawernawaz) [inline ecommerce/extensions/api/v2/tests/views/test_orders.py]: Initially I had tried to use `RequestFactory` but `ProductSerializer` return different url from [here](https://github.com/edx/ecommerce/blob/master/ecommerce/extensions/api/serializers.py#L149) what we have in response. Our response contains `testserver.fake` as domain name coming from here.
Using `RequestFactory` how I can override domain name.

Comment (clintonb) [inline ecommerce/extensions/api/v2/tests/views/test_orders.py]:
``` python
RequestFactory(SERVER_NAME=self.site.domain).get('/')
```

**Comment by tasawernawaz** [comment]
@clintonb I am continuously getting this error on build while running `make validate_js`
`Error: Cannot find module 'gulp-util`. Is this particular related to this PR.

**Comment by clintonb** [comment]
@tasawernawaz given that you aren't dealing with JavaScript, the issue is most likely unrelated. However, since it's preventing you from finishing this, you should investigate.
